Add support for LTC7871

adiceline opened the pull request:
## PR Description

- This adds driver support and device tree bindings for LTC7871 Six-Phase 
Buck-Boost Converter.
- Datasheet: [LTC7871](https://www.analog.com/media/en/technical-documentation/data-sheets/ltc7871.pdf)
- Tested on RPI4 using the following demo board: [DC2886A](https://www.analog.com/media/en/technical-documentation/user-guides/dc2886a.pdf)

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [X] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [X] I have conducted a self-review of my own code changes
- [X] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

adiceline committed:
dt-bindings: regulator: Document the ltc7871 regulator

The LTC7871 is a bidirectional buck or boost
switching regulator controller that operates in
either buck or boost mode on demand. Add
corresponding DT bindings.

Signed-off-by: Celine Joy A. Capua <celinejoy.capua@analog.com>

adiceline committed:
regulator: ltc7871: Add driver for LTC7871

Add ADI LTC7871 buck-boost controller driver support.

Signed-off-by: Celine Joy A. Capua <celinejoy.capua@analog.com>

adiceline committed:
kconfig.adi: Add LTC7871

imply REGULATOR_LTC7871.

Signed-off-by: Celine Joy A. Capua <celinejoy.capua@analog.com>

kister-jimenez commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+	int reg, ret;
+	int val1, val2;
+
+	chip->idac_setcur_uA = 16;
+	chip->freq_spread_range = 12;
+	chip->modulation_signal_freq = 512;
+	chip->enable_chip_ctrl_wp = 0;
+	chip->ra_ext = 10000;
+	chip->rb_ext = 107000;
+	chip->rc_ext = 12700;
+	chip->rd_ext = 499000;
```
Can you add a comment where you got these specific values?

kister-jimenez commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+
+	ret = ltc7871_reg_read(chip->spi, LTC7871_REG_CONFIG2, &reg);
+	if (ret < 0)
+		return ret;
+	chip->regulator_mode = FIELD_GET(LTC7871_CONFIG2_BUCK_BOOST, reg);
```
nit: add a newline after the return for better readability.

kister-jimenez commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+
+	ret = ltc7871_reg_read(chip->spi, LTC7871_REG_CHIP_CTRL, &reg);
+	if (ret < 0)
+		return ret;
+	chip->enable_chip_ctrl_wp = device_property_read_bool(&chip->spi->dev, "adi,enable-chip-ctrl-wp");
```
ditto.

kister-jimenez commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+	ret = device_property_read_u32(&chip->spi->dev, "adi,freq-spread-range",
+				       &chip->freq_spread_range);
+	if (!ret) {
+		ret = ltc7871_get_prop_index(ltc7871_freq_spread_range,
+					    ARRAY_SIZE(ltc7871_freq_spread_range),
+					    chip->freq_spread_range);
```
Align 2nd and 3rd argument to the first argument.

kister-jimenez commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+	}
+	ret = device_property_read_u32(&chip->spi->dev, "adi,modulation-signal-freq", &chip->modulation_signal_freq);
+	if (!ret) {
+		ret = ltc7871_get_prop_index(ltc7871_modulation_signal_freq,
+					    ARRAY_SIZE(ltc7871_modulation_signal_freq),
+					    chip->modulation_signal_freq);
```
ditto.

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,ltc7871-regulator.yaml:
```diff
@@ -0,0 +1,113 @@
+
+  adi,enable-chip-ctrl-wp:
+    description: Enables write for all three IDAC registers, and MFR_SSFM register.
+    type: boolean
```
Does `wp` stands for write protect? If so, the description should be improved

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,ltc7871-regulator.yaml:
```diff
@@ -0,0 +1,113 @@
+      15,
+      10,
+      8
+    ]
```
Do instead `enum: [12, 15, 10, 8]`. But more important is that the description does say much about what is this property about? Can we also express this in hz (or similar) and do the proper associations in the driver ?

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,ltc7871-regulator.yaml:
```diff
@@ -0,0 +1,113 @@
+      256,
+      128,
+      64
+    ]
```
Same comment as `adi,freq-spread-range`

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,ltc7871-regulator.yaml:
```diff
@@ -0,0 +1,113 @@
+            adi,freq-spread-range = <12>;
+            adi,modulation-signal-freq = <512>;
+          };
+        };
```
The indentation does not look coherent. Preferred is either 2 or 4 spaces... Also improve the commit title and messae. For the title `document the ltc7871 regulator` should be enough. For the message add a small description of the device. `Add documentation for devicetree bindings for LTC7871.` is just stating the obvious :)

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+
+static struct spi_driver ltc7871_driver = {
+	.driver = {
+		.name = "ltc7871"
```
please add an of_device_id table.

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+
+	chip->spi = spi;
+
+	return ltc7871_parse_fw(chip);
```
No call to [devm_regulator_register()](https://elixir.bootlin.com/linux/v6.12.6/C/ident/devm_regulator_register)?!

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+	device_property_read_u32(&chip->spi->dev, "adi,ra-external-ohm", &chip->ra_ext);
+	device_property_read_u32(&chip->spi->dev, "adi,rb-external-ohm", &chip->rb_ext);
+	device_property_read_u32(&chip->spi->dev, "adi,rc-external-ohm", &chip->rc_ext);
+	device_property_read_u32(&chip->spi->dev, "adi,rd-external-ohm", &chip->rd_ext);
```
Do we have any constrains for these values? Can we get any value from the user? I guess at least 0 should not be allowed?

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+	if (ret < 0)
+		return ret;
+	chip->enable_chip_ctrl_wp = device_property_read_bool(&chip->spi->dev, "adi,enable-chip-ctrl-wp");
+	val1 = (FIELD_PREP(LTC7871_MASK_CHIP_CTRL_WP, chip->freq_spread_range)) | reg;
```
no need for ()

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+	if (ret)
+		return ret;
+
+	device_property_read_u32(&chip->spi->dev, "adi,idac-setcur-microamp", &chip->idac_setcur_uA);
```
this looks to have more than 80 columns. Note that some maintainer still prefer this limit unless readability is hurt by the line break.

Also same comment about constrains. Can we get any value here?

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+					    chip->modulation_signal_freq);
+		if (ret < 0)
+			return ret;
+	}
```
nit: new line here

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+	struct ltc7871 *ltc7871 = rdev_get_drvdata(rdev);
+
+	if (sel < ltc7871->min_vol || sel > ltc7871->max_vol)
+		return -EINVAL;
```
In theory this should be controlled by the core AFAIR... Regulator constrains like this can be specified though DT and will be parsed during `of_get_regulator_init_data()` so I wonder if we need those extra properties for the resistors. IIUC, these resistors might affect minimum/max allowed voltages? Then, users could just properly set the regulator constrains properties depending on the resistors on their designs...

nunojsa replied:
Hmm, I see you still need the resistor values for `_ltc7871_uV_to_dac()` and `_ltc7871_dac_to_uV`.

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+	if (ret)
+		return ret;
+
+	*flags = 0;
```
Do we need the above?

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+	ret = spi_sync_transfer(spi, &t, 1);
+	if (ret < 0)
+		return ret;
+	return 0;
```
just `return spi_sync_transfer()`

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+
+	ret = spi_sync_transfer(spi, &t, 1);
+	if (ret < 0)
+		return ret;
```
nit: new line

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+	tmp = div_s64(tmp, r2);
+
+	return tmp;
+}
```
Just pass your `struct ltc7871 *ltc7871` and then you only need one argument. Also, now for sure that the resistors values cannot be 0

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+{
+	s64 tmp;
+
+	tmp = (1200 * (1000 + (div_s64(r2 * 1000, r1)))) - uV;
```
this looks to have more () than needed

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,342 @@
+
+static s64 _ltc7871_dac_to_uV(u32 dac_val, int r1, int r2)
+{
+	return (1200 * (1000 + (div_s64(r2 * 1000, r1)))) - (dac_val * r2);
```
ditto

adiceline commented:
### V2

**DT Bindings**

- Updated commit message
- Updated ` enable-chip-ctrl-wp` `description`
- Renamed `freq-spread-range` to `freq-spread-percentage`, changed `type`  to `string`, and updated `description`
- Renamed `modulation-signal-freq` to `switching-freq-divider` and updated description

**Kconfig**

- Updated Kconfig to reflect that LTC7871depends on both SPI and OF

**Driver**

- Added comment for default values
- Added `of_device_id`
- Added `devm_regulator_register`
- Added contraints for `ra_ext`, `rb_ext`, `rc_ext`, `rd_ext`, and `idac_setcur_uA`
- Changed `ltc7871_freq_spread_percentage` from `unsigned int` to `char *`
- Used `struct ltc7871 *ltc7871` as argument for `_ltc7871_uV_to_dac` and `_ltc7871_dac_to_uV` instead of `r1` and `r2`
- Fixed formatting issues

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,389 @@
+	if (IS_ERR(chip->rdev))
+		return PTR_ERR(chip->rdev);
+
+	return 0;
```
return `PTR_ERR_OR_ZERO()`

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,389 @@
+
+	chip = devm_kzalloc(dev, sizeof(*chip), GFP_KERNEL);
+	if (!chip)
+		return -ENOMEM;
```
nit: new line

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,389 @@
+		return ret;
+
+	device_property_read_u32(&chip->spi->dev, "adi,idac-setcur-microamp",
+				 &chip->idac_setcur_uA);
```
no error checking

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,389 @@
+	chip->rc_ext = 12700;
+	chip->rd_ext = 499000;
+
+	device_property_read_u32(&chip->spi->dev, "adi,ra-external-ohm",
```
add proper error checking on the calls to `device_property_read_u32()`. For non mandatory properties you can ignore errors but you still need to validate the property when given. Typical pattern is:

```
ret = device_property_read_u32(...);
if (!ret) {
    /* validate the propterty */
}
```

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,389 @@
+
+static s64 _ltc7871_dac_to_uV(struct ltc7871 *ltc7871, u32 dac_val)
+{
+	return 1200 * (1000 + (div_s64(ltc7871->r2 * 1000, ltc7871->r1))) - dac_val * ltc7871->r2;
```
looks like a longline

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,389 @@
+		return ret;
+
+	if (rx_buf[LTC7871_CRC_POS] != crc8(ltc7871_crc8_table, rx_buf,
+				LTC7871_CRC_POS, LTC7871_CRC_INIT))
```
maybe use some helper variables to make this more readable

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,389 @@
+#include <linux/regulator/driver.h>
+#include <linux/regulator/machine.h>
+#include <linux/regulator/of_regulator.h>
+#include <linux/bitfield.h>
```
headers should come in alphabetical order

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,ltc7871-regulator.yaml:
```diff
@@ -0,0 +1,102 @@
+  adi,ra-external-ohm:
+    description: External Resistor used to compute the value of
+      programmable VLOW in buck mode.
+    $ref: /schemas/types.yaml#/definitions/uint32
```
no need for the above when standard suffixes are used... Isn't ohms standard?

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,ltc7871-regulator.yaml:
```diff
@@ -0,0 +1,102 @@
+      divided to get the modulation frequency.
+    $ref: /schemas/types.yaml#/definitions/uint32
+    enum: [512, 1024, 2048, 4096, 256, 128, 64]
+    default: 512
```
can we somehow express this in frequency and do the math in the driver to get the actual divider?

adiceline replied:
The switching frequency is set via another external resistor which is arbitrary which means that the modulation frequency cannot be computed without prior knowledge of this resistor. Thus, we cannot directly use the modulation frequency as a property because it doesn't have fixed values.

adiceline replied:
An alternative would be to make the user input an integer for the modulation frequency, do the math in the driver, then round to the nearest valid value for the divisor. However, if we do this, the user will not be able to readback the actual value of the modulation frequency.

nunojsa replied:
> However, if we do this, the user will not be able to readback the actual value of the modulation frequency.

Why not? But for the above to work you would need to know the controller switching frequency so that you would need another resistor property right? I mean, if an external clock is used on the SYNC pin, it would be trivial to get the frequency but if you have the resistor on the FREQ pin, you would need to know it in order to do the math. Maybe let's go with the divider and see if upstream accepts it. Divider properties are not that uncommon after all.

But one thing though... you're missing an optional clock property for the SYNC pin when an external clock is used (should also be enabled in the driver). You could also add some regulator properties for the chip supplies.

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,402 @@
+				return ret;
+		} else {
+			return -EINVAL;
+		}
```
Typical pattern is to check for errors first... Flip the logic:

```
if (chip->idac_setcur_uA < LTC7871_IDAC_MIN ||
    chip->idac_setcur_uA > LTC7871_IDAC_MAX) {
    return dev_err_probe();

ret = ...
if (ret)
    return ret;
```

Makes the code neater. A log message with dev_err_probe() would also be nice so we know why we failed to probe (ditto for the other properties)

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,402 @@
+	ret = device_property_read_u32(&chip->spi->dev, "adi,ra-external-ohms",
+				 &chip->ra_ext);
+	if (!ret) {
+		if (chip->ra_ext <= 0)
```
Just `if (!chip->ra_ext)`. It's an unsigned value so checking for negative values is meaningless...

nunojsa commented on drivers/regulator/ltc7871-regulator.c:
```diff
@@ -0,0 +1,402 @@
+static s64 _ltc7871_dac_to_uV(struct ltc7871 *ltc7871, u32 dac_val)
+{
+	return 1200 * (1000 + (div_s64(ltc7871->r2 * 1000, ltc7871->r1))) -
+	       dac_val * ltc7871->r2;
```
I would also add a tmp variable as in _ltc7871_dac_to_uV() to make the above a bit more readable

nunojsa reviewed (commented):
Seems fairly good to send upstream. Take my last notes and please send the patches...
